### PR TITLE
fix: emit trait names in moniker identifier

### PR DESCRIPTION
**Pull request by tjdevries**
Hi,

Not sure if this is the best fix for the problem, but I noticed that functions that part of a trait are not namespaced with the trait name, so this could lead to problems if you had a function and a trait-function both defined in the same module.

Thanks!

TJ

### CODE REVIEW

File: crates/ide/src/moniker.rs [hunk 1]
```diff
@@ -106,9 +106,21 @@ pub(crate) fn def_to_moniker(
...
+        Definition::Field(it) => path.push(it.parent_def(db).name(db)),
+        Definition::Function(it) => {
+            // Ensure that trait functions are properly namespaced with the trait name
+            if let Some(assoc) = it.as_assoc_item(db) {
+                let container = assoc.container(db);
+                if let AssocItemContainer::Trait(parent_trait) = container {
+                    path.push(parent_trait.name(db));
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (tjdevries) [reply crates/ide/src/moniker.rs]: Oh, good point, thanks. I think I got it with the latest commit. Added tests for each of the cases.

**Comment by tjdevries** [comment]
I think I'm not handling implementations though but I'm not sure how to get the name of the function that is getting implemented.

For example:

```rust
    #[test]
    fn moniker_for_trait_impl_function() {
        check_moniker(
            r#"
//- /foo/lib.rs crate:foo@CratesIo:0.1.0,https://a.b/foo.git
pub mod module {
    pub trait MyTrait {
        pub fn func() {}
    }

    struct MyStruct {}

    impl MyTrait for MyStruct {
        pub fn func$0() {}
    }
}
"#,
            "foo::module::MyStruct::func",
            r#"PackageInformation { name: "foo", repo: "https://a.b/foo.git", version: "0.1.0" }"#,
            MonikerKind::Export,
        );
    }
```

I'm not sure how to get the `MyStruct` from `impl MyTrait for MyStruct` in this case.

The other thing I'm wondering is should the moniker be: `foo::module::MyStruct::MyTrait::func` or `foo::module::MyStruct::func`.

Thanks

**Comment by Veykril** [comment]
You can fetch the trait of an impl be checking if the container is an `impl` and then querying the `self_ty`. So something like `impl_.self_ty(db).as_adt().expect("handle this").name(db)`. How the moniker should look like I can't tell you though as I have no knowledge about it.

**Pull request merged**